### ISSUE TITLE
Upgradable ERC20

### ISSUE THREAD

**Issue by barakman** [OpenZeppelin/openzeppelin-contracts#1499]
**🧐 Motivation**

We have a contract which inherits from `ERC20`, let's call it `MyToken`.
We need to allow ourselves to upgrade this contract in the future.
Due to the nature of `ERC20`, we cannot simply replace the `MyToken` instance with a new instance.
That would require pausing all functionalities and migrating all data (`_balances`, `_totalSupply`, etc).
Furthermore, it would wreck any promise for transparency, as we would have the power to delete all funds in an instant.

**📝 Details**

Ideally, we would like to do this as follows:
1. In `MyToken`, override each one of the main `ERC20` functions (`transfer` and `transferFrom`) to simply call its correspondent function in `MyTokenImpl`, and pass `msg.sender` to it as an additional argument.
2. In `MyTokenImpl`, in each function, call back a function in `MyToken` to perform the desired action.

For example:
```
interface IMyToken {
    function transferCallback(address from, address to, uint256 value) public;
    function transferFromCallback(address sender, address from, address to, uint256 value) public;
}

interface IMyTokenImpl {
    function transfer(address sender, address to, uint256 value) public returns (bool);
    function transferFrom(address sender, address from, address to, uint256 value) public returns (bool);
}

contract MyToken is IMyToken, ERC20, Ownable {
    IMyTokenImpl private pMyTokenImpl;
    function upgrade(IMyTokenImpl _pMyTokenImpl) public onlyOwner {
        pMyTokenImpl = _pMyTokenImpl;
    }
    function transfer(address to, uint256 value) public returns (bool) {
        return pMyTokenImpl.transfer(msg.sender, to, value);
    }
    function transferFrom(address from, address to, uint256 value) public returns (bool) {
        return pMyTokenImpl.transferFrom(msg.sender, from, to, value);
    }
    function transferCallback(address from, address to, uint256 value) public {
        require(msg.sender == address(pMyTokenImpl));
        _transfer(from, to, value);
    }
    function transferFromCallback(address sender, address from, address to, uint256 value) public {
        require(msg.sender == address(pMyTokenImpl));
        _allowed[from][sender] = _allowed[from][sender].sub(value);
        _transfer(from, to, value);
    }
}

contract MyTokenImpl is IMyTokenImpl {
    IMyToken private pMyToken;
    constructor(IMyToken _pMyToken) public {
        pMyToken = _pMyToken;
    }
    function transfer(address sender, address to, uint256 value) public returns (bool) {
        require(msg.sender == address(pMyToken));
        pMyToken.transferCallback(sender, to, value);
    }
    function transferFrom(address sender, address from, address to, uint256 value) public returns (bool) {
        require(msg.sender == address(pMyToken));
        pMyToken.transferFromCallback(sender, from, to, value);
    }
}
```
As you might understand, the problem is in:
```
_allowed[from][sender] = _allowed[from][sender].sub(value);
```
This is because the access-level of variable `_allowed` is `private`.

Would you be able to provide an `internal` function (similar to function `_transfer`, for example) to execute this line?

For example:
```
function _prohibit(address from, address to, uint256 value) internal {
    _allowed[from][to] = _allowed[from][to].sub(value);
}
```

Minor: by then, you may as well call it from function `transferFrom`.

Thanks

**Comment by nventuro** [comment]
Hi again @barakman!

What you're describing is one of the standard ways of implementing upgradeability. While we're considering adding that `internal` function, have you looked into off-the-shelf upgradeability solutions, such as [ZeppelinOS](https://zeppelinos.org/)? Upgradeability can be quite tricky, and it looks like you may be reinventing the wheel a bit in this case.

As a sidenote, the reason `_transfer` exists is not to encourage these sorts of mechanisms, but to allow for derived contracts to easily and safely manage balances (if needed).

**Comment by barakman** [comment]
@nventuro: Hi.

We have already established a very simple infrastructure which allows us to:
1. Restrict certain functions to certain contracts (e.g., function `func` in contract `A` can be invoked only from contract `B`).
2. Execute a partial or full system-upgrade (i.e., replace some or all contracts) as an atomic operation, while each contract remains "oblivious" to the replacement of any other contract in the system.

As I said, this infrastructure is very simple and small, which gives us better confidence of robustness and security, so we'd like to keep it this way (I can share if needed).

In any case, the only problem remaining is, that if we replace the ERC20 contract itself (with which users interact when they buy or sell tokens), then we need to declare the new contract address to the world.
This would be equivalent to a *Fork* on our system, and we'd like to avoid it as much as possible.

Hence, we've decided to split it into two contracts:
1. The ERC20 contract itself (our user-interface), which contains as little logic as possible.
2. A contract which implements each one of the supported operations, which we can replace without having to re-announce an entry point (i.e., a new contract address).

The replacement will be done via the mechanism that I have described above (i.e., system-upgrade).
Of course, this mechanism also allows us to replace the ERC20 contract itself if needed, but as I said, we'd like to minimize the chance of it ever happening.

Thanks

**Comment by spalladino** [comment]
Hey @barakman! What you are implementing is quite similar to what we're doing with upgradeability in ZeppelinOS: we have a proxy contract (equivalent to your `MyToken`) that holds the balances of the users, and whose address will never change (so end users can interact always with the same address). That proxy delegates execution to an ERC20 implementation (`MyTokenImpl` in your case). Do you mind checking out [our docs](https://docs.zeppelinos.org/docs/deploying.html), to see if it fits your needs? I think it may be a more robust solution than an ad-hoc token wrapper.

**Comment by barakman** [comment]
@spalladino:
Thank you for your quick response.
Can you please provide a link directly to the source code on GitHub?
Thanks

**Comment by spalladino** [comment]
Sure! It's on https://github.com/zeppelinos/zos

**Comment by barakman** [comment]
@spalladino:
Thank you!

On a quick glance, I see numerous contracts and a rather large amount of code in general.
Our upgradability infrastructure consists of only 3 contracts and just a handful of code lines.
So between the accountability of your infrastructure (tested and reviewed by the community) and the simplicity of ours, I currently tend to go with the latter.

Having said that, we are just at the beginning of an audit procedure with several providers, so I'll stand await for any suggestions that they may have.
We have already been advised to use OZ via `npm install` (instead of locally), and to upgrade from v1.12.0 to v2.0.0 (a day after it was officially released!), so I'm assuming that these auditers are aware of the alternatives (and aware of OZ in general).

Having said that 2, is there a one-to-one equivalence of my coding example above, that you can refer me to?
In other words, do you have a pair of contracts splitting an ERC20 contract into data and logic?
Thanks

**Comment by spalladino** [comment]
@barakman certainly. The ERC20 data would be managed by the [Proxy contract](https://github.com/zeppelinos/zos/blob/master/packages/lib/contracts/upgradeability/AdminUpgradeabilityProxy.sol), while the ERC20 used can be provided by the [OpenZeppelin EVM package](https://github.com/OpenZeppelin/openzeppelin-eth/blob/master/contracts/token/ERC20/StandaloneERC20.sol). Using those two contracts only you can reproduce the same setup.

**Comment by frangio** [comment]
You can find a list of resources about upgradeability in our documentation site.

https://docs.openzeppelin.com/openzeppelin/upgrades
